instrumentation/servlet/servlet-3.0/src/main/java/io/opentelemetry/auto/instrumentation/servlet/v3_0/CountingHttpServletResponse.java

```diff
@@ -0,0 +1,146 @@
+ * limitations under the License.
+ */
+
+package io.opentelemetry.auto.instrumentation.servlet.v3_0;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.Writer;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.servlet.ServletOutputStream;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+
+/** HttpServletResponseWrapper since servlet 2.3, not applicable to 2.2 */
+public class CountingHttpServletResponse extends HttpServletResponseWrapper {
+  private CountingServletOutputStream outputStream = null;
+  private CountingPrintWriter printWriter = null;
+  private int errorLength = 0;
+
+  /**
+   * Constructs a response adaptor wrapping the given response.
+   *
+   * @throws IllegalArgumentException if the response is null
+   */
+  public CountingHttpServletResponse(HttpServletResponse response) {
+    super(response);
+  }
+
+  @Override
+  public ServletOutputStream getOutputStream() throws IOException {
+    if (outputStream == null) {
+      outputStream = new CountingServletOutputStream(super.getOutputStream());
+    }
+    return outputStream;
+  }
+
+  @Override
+  public PrintWriter getWriter() throws IOException {
+    if (printWriter == null) {
+      printWriter = new CountingPrintWriter(super.getWriter());
+    }
+    return printWriter;
+  }
+
+  public int getContentLength() {
+    int contentLength = errorLength;
+    if (outputStream != null) {
+      contentLength += outputStream.counter;
+    }
+    if (printWriter != null) {
+      contentLength += printWriter.counter.get();
+    }
+    return contentLength;
+  }
+
+  /** sendError bypasses the servlet response writers and writes directly to the response */
+  @Override
+  public void sendError(int sc, String msg) throws IOException {
+    super.sendError(sc, msg);
+    if (msg != null) {
+      errorLength += msg.length();
+    }
+  }
+
+  static class CountingServletOutputStream extends ServletOutputStream {
+
+    private ServletOutputStream delegate;
+    private int counter = 0;
+
+    public CountingServletOutputStream(ServletOutputStream delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+      delegate.write(b);
+      counter++;
+    }
+
+    @Override
+    public void write(byte[] b) throws IOException {
+      delegate.write(b);
+      counter += b.length;
+    }
+
+    @Override
+    public void write(byte[] b, int off, int len) throws IOException {
+      delegate.write(b, off, len);
+      counter += len;
+    }
+
+    @Override
+    public void flush() throws IOException {
+      delegate.flush();
+    }
+
+    @Override
+    public void close() throws IOException {
+      delegate.close();
+    }
+  }
+
+  static class CountingPrintWriter extends PrintWriter {
+    // PrintWriter is synchronised, so the counter has to be atomic
+    private AtomicInteger counter = new AtomicInteger(0);
```

Comment:
The jdk implementation uses an internal lock to synchronise pretty much every method. Since I can't grab that lock I can imagine a situation under high load where two of those methods are called, pulled off of the cpu, then scheduled on two cpus at the same time.

Reply:
I think the Servlet Spec comes to our rescue here:

> Other than the startAsync and complete methods, implementations of the request
and response objects are not guaranteed to be thread safe. This means that they
should either only be used within the scope of the request handling thread or the
application must ensure that access to the request and response objects are thread
safe.

My reading of this is that we don't even need volatile here (similar to implementation above in `CountingServletOutputStream`).